### ISSUE TITLE
winpmem 3.1rc3 - dumping in "raw" format never completes on Windows, output file grows until no disk space left

### ISSUE THREAD

**Issue by justfalter** [Velocidex/c-aff4#40]
On Windows, attempts to use winpmem 3.1rc3 to dump memory in "raw" format never completes, and the output file created grows so large (hundreds of gigabytes in size) that the disk runs out of space. I've narrowed this down to only happening when the system has memory ranges with starting addresses >= 0x100000000 (2^32). When the memory ranges are confined to < 0x100000000, raw memory dumps succeed. 

On Windows 7, I found these ranges would start showing up when I configured the system to have >= 3584MB (3.5GB) of RAM. At 3583MB of RAM, raw memory dumping would succeed. 

The command I'm using in all cases is:
```
winpmem_3.1.rc3.exe --format raw --output raw.dump
```

Here are runs that illustrate the problem that I am encountering
- Windows 7 64 bit, 3583MB of RAM:
  - Memory range start addresses: 0x00001000, 0x00100000
   - raw memory dump completes
   - [Command output](https://gist.github.com/justfalter/bc5009124043fa4609f5602720cd713d#file-windows7-3583mb-ram)
- Windows 7  64 bit, 3584MB of RAM
  - Memory range start addresses: 0x00001000, 0x00100000, 0x100000000
  - This seems to be the magic amount of memory on Windows 7 for the 0x100000000 memory address to start being used.
  - raw memory dump never completes, output file is hundreds of gigabytes in size, no disk space left.
  - [Command output](https://gist.github.com/justfalter/bc5009124043fa4609f5602720cd713d#file-windows7-3584mb-ram)
- Windows 7 64 bit, 4096MB of RAM
  - Memory range start addresses: 0x00001000, 0x00100000, 0x100000000
  - raw memory dump never completes, output file is hundreds of gigabytes in size, no disk space left.
  - [Command output](https://gist.github.com/justfalter/bc5009124043fa4609f5602720cd713d#file-windows-7-4gb-ram)
- Windows 10 64 bit, 16GB of RAM
  - Memory range start addresses: 0x00001000, 0x00100000, 0x00103000, 0x100000000
  - raw memory dump never completes, output file is hundreds of gigabytes in size, no disk space left.
  - [Command output](https://gist.github.com/justfalter/bc5009124043fa4609f5602720cd713d#file-windows-10-16gb-of-ram)

**Comment by justfalter** [comment]
FWIW, I see the same behavior with winpmem_3.0.rc3.exe and winpmem_3.0.rc1.exe.

**Comment by justfalter** [comment]
I've been doing some additional debugging, and I believe that I may have narrowed things down a bit. I'm using [API Monitor](http://www.rohitab.com/apimonitor) to track CreateFile, WriteFile, ReadFile, and SetFilePointerEx API calls in winpmem_3.1.rc3.exe. 

On my Windows 10 machines with 16GB of RAM, I have the following memory ranges:
```
Start 0x00001000 - Length 0x0009E000
Start 0x00100000 - Length 0x00002000
Start 0x00103000 - Length 0xBFED7000
Start 0x100000000 - Length 0x340000000
```

Our final raw memory dump should have the following construction. I'm naming the regions, as this will help me out, later. 
```
0x00000000 - 0x00000FFF   - Null padding 0
0x00001000 - 0x0009EFFF    - Memory Region 1
0x0009F000 - 0x000FFFFF    - Null padding 1
0x00100000 - 0x00101FFF     - Memory Region 2
0x00102000 - 0x00102FFF     - Null padding 2
0x00103000 - 0xBFFD9FFF    - Memory Region 3
0xBFFDA000 - 0xFFFFFFFF    - Null padding 3
0x100000000 - 0x43FFFFFFF   - Memory Region 4
```

Here's a common sequence for a cycle of reading from memory, writing to a dump:
- `SetFilePointerEx(readFileHandle, memoryOffsetToStartReadingFrom, &lpNewFilePointer, FILE_BEGIN)`
- And then a normal ReadFile operation for some amount of data (usually 32768 bytes)
- `SetFilePointerEx(readFileHandle, offsetIntoDumpFile, &writeOffset, FILE_BEGIN)`
- And then a WriteFile operation for the data.

With this, I'm able to monitor which memory regions are being read from by tracking the SetFilePointerEx calls for the read file handle. 

From my tracing, it appears that the memory dump that's being sent to file looks like the following is a repeating pattern:
- Null padding 0
- Memory Region 1
- Null padding 1
- Memory Region 2
- Null Padding 2
- Memory Region 3
- Null Padding 3
- Null padding 0
- Memory Region 1
- Null padding 1
- Memory Region 2
- Null Padding 2
- Memory Region 3
- Null Padding 3
- Null padding 0
- Memory Region 1
- Null padding 1
- Memory Region 2
- Null Padding 2
- Memory Region 3
- Null Padding 3
- ... and so on

It appears that winpmem is never moving on to actually write Memory Region 4, and is somehow stuck in a loop, repeating the writing of all but the last memory region. I suspect that the issue lies in AFF4Map::Read, specifically in this area: https://github.com/Velocidex/c-aff4/blob/master/aff4/aff4_map.cc#L106-L121

**Comment by scudette** [comment]
Thanks for the detailed analysis! Please see the latest pre-release binary for a fix.

**Comment by justfalter** [comment]
@scudette Just tested out 3.1 rc5 on my Windows 10 VM. While it no longer dumps into infinity, it now seems to hang after writing exactly 5GB of data. I tested this both with the ".dump" and ".aff4" file extensions, and there's no difference. 
Here are gists for you:
https://gist.github.com/justfalter/087556871bde0cf097151c4f040f7ee4

**Comment by justfalter** [comment]
I think you've got another integer overflow right here: https://github.com/Velocidex/c-aff4/blob/d1deb8f486c27cbc8d01153224572a2c5479aada/aff4/aff4_map.cc#L129
```
        size_t length_to_read_in_target = std::min(
            length, (size_t)(range.map_end() - readptr));
```

- range.map_end() is a uint64_t
- readptr is an aff4_off_t, which is an int64_t. 
- and then you're casting it to a size_t, which is a uint32_t.

I don't really understand why BinaryRange would have map_offset defined as uint64_t, when it seems like they should be aff4_off_t. Is there any particular reason?  https://github.com/Velocidex/c-aff4/blob/master/aff4/aff4_map.h#L31-L36

I might be able to come up with a PR to fix this, but I'm short on time at the moment.

**Comment by justfalter** [comment]
Alright, I've got a fix in #44
